### PR TITLE
Wait for events when managing on-chain services

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR

- Updates the `autonomy service register` command to take multiple agent instances as input at a time
- Introduces waiting mechanism for various events
- Introduces `--timeout` flag to provide custom timeout value to wait for the on-chain events

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: autonomy/chain/mint.py [hunk 1]
```diff
@@ -75,13 +78,32 @@ def sort_service_dependency_metadata(
...
+
+    timeout = timeout or EVENT_VERIFICATION_TIMEOUT
+    deadline = datetime.datetime.now() + datetime.timedelta(seconds=timeout)
+    while datetime.datetime.now() < deadline:
+        token_id = token_retriever()
+        if token_id is not None:
+            return token_id
+        time.sleep(sleep)
+
+    raise TimeoutError("Could not retrieve the token in given time limit.")
```
Comment (angrybayblade) [inline autonomy/chain/mint.py:96]: from now we will check for the events multiple times in a given time limit

File: autonomy/chain/service.py [hunk 2]
```diff
@@ -89,11 +92,63 @@ def get_service_info(
     )
 
 
+def wait_for_success_event(
```
Comment (angrybayblade) [inline autonomy/chain/service.py:95]: In the same manner we will also wait for the success event for various service operations

File: autonomy/cli/mint.py [hunk 3]
```diff
@@ -57,21 +57,26 @@
     type=str,
     help="IPFS hash for the NFT image",
 )
+timeout_flag = click.option(
+    "-t", "--timeout", type=float, help="Timeout for verifying emnitted events"
+)
```
Comment (angrybayblade) [inline autonomy/cli/mint.py:62]: To customise the event verification timeout

File: autonomy/cli/service.py [hunk 4]
```diff
@@ -72,36 +75,41 @@ def activate(
 @click.option(
     "-i",
     "--instance",
+    "instances",
     type=str,
     required=True,
+    multiple=True,
```
Comment (angrybayblade) [inline autonomy/cli/service.py:81]: This was the reason why all 4 agent instances were not getting registered at the same time, although we were providing 4 agent instances the command was only taking one of them as input, this was my mistake on the design part

**Review by kupermind — APPROVED** [review]

**Review by solarw — APPROVED** [review]
lgtm

**Review by DavidMinarsch — APPROVED** [review]

File: autonomy/chain/constants.py [hunk 5]
```diff
@@ -29,6 +29,8 @@
 from autonomy.data import DATA_DIR
 
 
+EVENT_VERIFICATION_TIMEOUT = 60.0
```
Comment (DavidMinarsch) [inline autonomy/chain/constants.py:32]: In another PR suggest bumping this to 180.0 - 60.0 is a bit low

**Pull request merged**